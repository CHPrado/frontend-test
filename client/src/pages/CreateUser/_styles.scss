#page-create-user {
  width: 100%;
  max-width: 1100px;
  margin: 0 auto;
  height: 100vh;

  @extend .flex;

  form {
    margin: 40px auto;
    padding: 30px;
    max-width: 730px;
    background: #FFF;
    border-radius: 8px;
  
    @extend .flex;
    flex-direction: column;

    h1 {
      font-size: 35px;
      color: $title-color;
      margin-bottom: 20px;
    }

    .avatar {
      @extend .flex;
      flex-direction: column;
      border-radius: 10rem;
        width: 300px;
    
      margin-top: 12px;
      margin-bottom: 12px;
      cursor: pointer;
      
      img {
        border-radius: 10rem;
        width: 300px;
      }
      
      img:hover {
        filter: brightness(90%);
      }
      
      strong {
        margin: 5px;
        color: #737380;
        line-height: 21px;
        font-size: 16px;
      }
    }

    .field-group {
      width: 100%;
      @extend .flex;

      .field {
        width: 100%;
        margin-bottom: 24px;

        @extend .flex;
        flex-direction: column;
        align-items: flex-start;

        + .field {
          margin-left: 24px;
        }

        input {
          flex: 1;
          width: 100%;
          padding: 16px 24px;
          font-size: 16px;
          border-radius: 8px;
          border: 2px solid #A0A0A9;
          color: $text-color;

          &::placeholder {
            color: #A0A0A9;
          }

          &:focus {
            outline: none !important;
            border: 2px solid $primary-color;
            box-shadow: 0 0 5px #719ECE;
          }
        }
      }
    }

    button {
      width: 100%;
      height: 50px;
      background: $primary-color;
      border-radius: 8px;
      color: #FFF;
      font-weight: bold;
      font-size: 16px;
      border: 0;
      align-self: center;
      margin-top: 20px;
      transition: background-color 0.2s;
      cursor: pointer;

      &:hover {
        background: $button-hover-color;
      }
    }
  }
}

@media(max-width: 900px) {
  #page-create-user form .field-group {
    flex-direction: column;

    .field + .field {
      margin-left: 0;
    }
  }
}