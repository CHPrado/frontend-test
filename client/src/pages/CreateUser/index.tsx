import React, { ChangeEvent, useState, FormEvent, useEffect } from 'react';
import { useHistory } from 'react-router-dom';

import api from '../../services/api';
import Snack, { SnackProps } from '../../components/Snack';
import portrait from '../../assets/portrait.png';

const CreateUser = () => {
  const [avatar, setAvatar] = useState<string>(portrait);
  const [formData, setFormData] = useState({
    name: '',
    username: '',
    email: '',
    phone: '',
  });

  const history = useHistory();

  const [alertState, setAlertState] = useState(false);
  const [alertType, setAlertType] = useState<SnackProps['type']>('');
  const [alertMessage, setAlertMessage] = useState('');

  let inputPic: HTMLInputElement;

  const [file, setFile] = useState<File>();

  useEffect(() => {
    if (alertType === 'success') {
      if (!alertState) history.push('/');
    }
  }, [alertState]); // eslint-disable-line react-hooks/exhaustive-deps

  useEffect(() => {
    if (file) {
      const reader = new FileReader();
      reader.readAsDataURL(file);
      reader.onload = (event) => {
        setAvatar(event?.target?.result as string);
      };
    }
  }, [file]);

  function getRef(input: HTMLInputElement) { inputPic = input; };

  function handlePicClick() {
    inputPic.click();
  };

  function handleCloseAlert() {
    setAlertState(false);
  };

  function handleInputChange(event: ChangeEvent<HTMLInputElement>) {
    const { name, value } = event.target;

    setFormData({ ...formData, [name]: value });
  }

  function stopSubmit() {
    let msg = '';
    const { name, username, email, phone } = formData;

    if (name === '') msg = 'Preencha o campo Nome';
    else if (username === '') msg = 'Preencha o campo Nome de Usuário';
    else if (email === '') msg = 'Preencha o campo E-Mail';
    else if (phone === '') msg = 'Preencha o campo Telefone';
    else if (avatar === portrait) msg = 'Adicione uma Foto';

    if (msg !== '') {
      setAlertMessage(msg);
      setAlertType('warn');
      setAlertState(true);
      return true;
    }

    return false;
  }

  async function handleSubmit(event: FormEvent) {
    event.preventDefault();

    const { name, username, email, phone } = formData;

    const data = {
      name,
      username,
      email,
      phone,
      avatar
    };

    if (stopSubmit()) return;

    await api.post('users', data).then(() => {
      setAlertMessage('Usuário criado! Você será redirecionado.');
      setAlertType('success');
      setAlertState(true);

      window.setTimeout(() => {
        history.push('/');
      }, 2000);
    });
  }

  return (
    <>
      <Snack
        type={alertType}
        open={alertState}
        close={handleCloseAlert}
        message={alertMessage}
      />
      <div id="page-create-user">
        <form onSubmit={handleSubmit}>
          <h1>Cadastro de usuário</h1>

          <div
            className="avatar"
            onClick={handlePicClick}
          >
            <img src={avatar} alt="avatar" />
            <strong>Adicione uma Foto</strong>

            <input
              hidden
              type="file"
              accept="image/*"
              ref={getRef}
              onChange={(e) => {
                setFile(e?.target?.files ? e?.target?.files[0] : undefined);
              }}
            />
          </div>

          <div className="field-group">
            <div className="field">
              <input
                type="text"
                name="name"
                id="name"
                placeholder="Nome"
                onChange={handleInputChange}
              />
            </div>
            <div className="field">
              <input
                type="text"
                name="username"
                id="username"
                placeholder="Nome de Usuário"
                onChange={handleInputChange}
              />
            </div>
          </div>

          <div className="field-group">
            <div className="field">
              <input
                type="text"
                name="email"
                id="email"
                placeholder="E-Mail"
                onChange={handleInputChange}
              />
            </div>
            <div className="field">
              <input
                type="text"
                name="phone"
                id="phone"
                placeholder="Telefone"
                onChange={handleInputChange}
              />
            </div>
          </div>

          <button type="submit">
            Cadastrar usuário
          </button>
        </form>
      </div>
    </>
  );
};

export default CreateUser;
