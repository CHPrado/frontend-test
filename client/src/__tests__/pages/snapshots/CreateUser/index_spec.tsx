import React from 'react';
import renderer from 'react-test-renderer';

import CreateUser from '../../../../pages/CreateUser';

test('it matches the snapshot', () => {
  const tree = renderer.create(<CreateUser />).toJSON();

  expect(tree).toMatchSnapshot();
});
