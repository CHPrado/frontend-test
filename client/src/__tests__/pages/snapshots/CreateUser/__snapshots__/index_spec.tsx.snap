// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches the snapshot 1`] = `
Array [
  <div
    className="snack-container hide"
  >
    <div
      className="snack-box "
    >
      <div
        className="snack-icon"
      />
      <div
        className="snack-message"
      >
        <strong>
          
        </strong>
      </div>
      <div
        className="snack-close "
        onClick={[Function]}
        role="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </div>
    </div>
  </div>,
  <div
    id="page-create-user"
  >
    <form
      onSubmit={[Function]}
    >
      <h1>
        Cadastro de usuário
      </h1>
      <div
        className="avatar"
        onClick={[Function]}
      >
        <img
          alt="avatar"
          src="portrait.png"
        />
        <strong>
          Adicione uma Foto
        </strong>
        <input
          accept="image/*"
          hidden={true}
          onChange={[Function]}
          type="file"
        />
      </div>
      <div
        className="field-group"
      >
        <div
          className="field"
        >
          <input
            id="name"
            name="name"
            onChange={[Function]}
            placeholder="Nome"
            type="text"
          />
        </div>
        <div
          className="field"
        >
          <input
            id="username"
            name="username"
            onChange={[Function]}
            placeholder="Nome de Usuário"
            type="text"
          />
        </div>
      </div>
      <div
        className="field-group"
      >
        <div
          className="field"
        >
          <input
            id="email"
            name="email"
            onChange={[Function]}
            placeholder="E-Mail"
            type="text"
          />
        </div>
        <div
          className="field"
        >
          <input
            id="phone"
            name="phone"
            onChange={[Function]}
            placeholder="Telefone"
            type="text"
          />
        </div>
      </div>
      <button
        type="submit"
      >
        Cadastrar usuário
      </button>
    </form>
  </div>,
]
`;
